### DISCUSSION TITLE
在msys下的ucrt64中如何使用gcc工具链

### DISCUSSION THREAD

**Discussion by supf1994** [xmake-io/xmake#4269] [Q&A]
在ucrt64中默认检测到的编译器是vs，使用 xmake f --toolchain=gcc 也不生效
使用的xmake版本是2.8.3

**Comment by star-hengxing** [comment]
先切换平台
`xmake f -p msys`

**Comment by supf1994** [comment]
不行，检查出来还是  vs
![image](https://github.com/xmake-io/xmake/assets/23397492/4e10eb22-eabb-40c9-8ff7-e66bae2313e4)

**Reply by supf1994** [reply]
我是在mingw中直接使用pacman安装xmake，跟这个有关系吗，我把msys整个删除重装，然后再使用pacman 在mingw64上安装，也是一样的问题

**Reply by star-hengxing** [reply]
Windows 上默认检查 vs，不影响编译

**Reply by waruqi** [reply]
提供 xmake.lua 和完整 -vD 输出

**Reply by supf1994** [reply]
`add_rules("mode.debug", "mode.release")

set_languages("c11", "cxx20")
add_requires("openssl")
add_packages("openssl")

if is_mode("debug") then 
    add_defines('_DEBUG')
end 


-- window下的特殊处理
if is_plat("windows") then 
    if is_mode("debug") then 
        set_runtimes("MDd")     --使用vld时必须设置为MDd
    else 
        set_runtimes("MD")
    end 
    
    -- 添加boost库相关路径
    add_includedirs("D:/Develop/Lib/Vld/include")
    add_includedirs("D:/Develop/Lib/Boost/boost_1_83_0_build/include")
    if is_arch("x86") then 
        add_linkdirs("D:/Develop/Lib/Vld/x86/lib")
        add_linkdirs("D:/Develop/Lib/Boost/boost_1_83_0_build/x32/lib")
    elseif is_arch("x64") then 
        add_linkdirs("D:/Develop/Lib/Vld/x64/lib")
        add_linkdirs("D:/Develop/Lib/Boost/boost_1_83_0_build/x64/lib")
    else
        error("windows不支持的CPU架构")
    end

    -- 支持大的编译目标文件
    add_cxflags("/bigobj")

    -- 设置MSVC为UTF8编码
    add_cxflags("/utf-8")
    -- vs下nodiscard标记的返回值必须被处理，减少因协程忘记co_await导致的bug
    add_cxflags("/we4858", {force = true})
    --add_cxxflags("/we4834", {force = true})
    
else 
    --add_requires("boost")     -- 添加boost库支持
    if is_plat("mingw") then 
        add_links("wsock32")
        --set_policy("package.include_external_headers", false)
    end
end


includes("tools")
includes("LL-Device-Server")`

-> tools/xmake.lua
`target("tools")
    set_kind("static")

    add_rules("c++.unity_build", {batchsize = 8})

    add_includedirs("./")
    add_headerfiles("**.h", "**.hpp")
    add_files("**.cpp", "**.c")

    if not is_plat("windows") then 
        add_links("boost_locale-mt", "boost_filesystem-mt")
    end`

-> LL-Device-Server/xmake.lua
`add_requires("toml++")
if not is_plat("mingw") then 
    add_requires("pugixml")
end


target("LL-Device-Server")
    set_kind("binary")
    if is_plat("windows") then 
        set_pcxxheader("entry/common_header.h")      --msvc下预编译头速度快
    else 
        add_rules("c++.unity_build")      --gcc使用预编译头每次改动都得整个项目重新编译，因此不使用预编译头，采用unity_build
    end
    add_deps("tools")
    add_packages("toml++")
    if not is_plat("mingw") then 
        add_packages( "pugixml")
    else
        add_links("pugixml")
    end

    add_includedirs("./", "../", "../tools")
    add_headerfiles("**.h", "**.hpp")
    add_files("**.cpp", "**.c")
    add_files("global/**.cpp", {unity_group = "global"})
    add_files("entry/**.cpp", {unity_ignored = true})
    --add_files("server/**.cpp", {unity_ignored = true})
    add_files("server/**.cpp", {unity_group = "server"})`

![image](https://github.com/xmake-io/xmake/assets/23397492/ea109524-e316-4518-ad94-7047132ef0c3)

**Reply by waruqi** [reply]
这不挺正常么，没看出什么问题

**Reply by supf1994** [reply]
好的，是我看错了，看到检查vs版本就以为用vs来编译，不过在ucrt64上面平台检查应该是mingw吧
